File: checkpoint0/src/main/java/com/springernature/checkpoint0/BoardRenderer.java
package com.springernature.checkpoint0;

import java.util.stream.IntStream;
import static java.util.stream.Collectors.joining;

class BoardRenderer {
    private final Board board;

    BoardRenderer(Board board) {
        this.board = board;
    }

    String render() {
        return IntStream.range(0, board.size())
                .mapToObj(this::renderRow)
                .collect(joining("-----\n"));
    }

    private String renderRow(int row) {
        return IntStream.range(0, board.size())
                .mapToObj(column -> renderCell(row, column) + squareSeparator(column))
                .collect(joining()) + "\n";
    }

    private String renderCell(int row, int column) {
        switch (board.get(row, column)) {
            case X: return "X";
            case O: return "O";
            case Empty: return " ";
            default: throw new IllegalStateException("It never happens!");
        }
    }

    private String squareSeparator(int column) {
        return column < board.size() - 1 ? "|" : "";
    }
}

File: checkpoint1/src/main/kotlin/com/springernature/checkpoint1/BoardRenderer.kt
package com.springernature.checkpoint1

import com.springernature.checkpoint1.Cell.*

class BoardRenderer(private val board: Board) {

    fun render() =
            0.until(board.size)
                    .map { this.renderRow(it) }
                    .joinToString("-----\n")

    private fun renderRow(row: Int) =
            0.until(board.size)
                    .map { column -> renderCell(row, column) + squareSeparator(column) }
                    .joinToString("") + "\n"

    private fun renderCell(row: Int, column: Int) =
            when (board[row, column]) {
                X -> "X"
                O -> "O"
                Empty -> " "
            }

    private fun squareSeparator(column: Int) = if (column < board.size - 1) "|" else ""
}
